Uplift tt-llk to include high-precision reciprocal, sqrt, and rsqrt.

jasondavies opened the pull request:
Uplift tt-llk to include high-precision reciprocal, sqrt, and rsqrt.

As part of this change:

- A new legacy_compat template parameter to had to be added to recip_tile, sqrt_tile, and rsqrt_tile to temporarily keep the old implementations until various users such as layernorm fix their bugs.
- Uses of sqrt_tile+recip_tile are replaced with rsqrt_tile where possible.
- Hardcoded constants in users of exp/recip LLKs are replaced with the appropriate _init functions.
- rsqrt_tile has been changed to use `approx_mode` instead of `fast_and_approx_mode` (which was completely ignored).
- Dropped `RECIPROCAL_ITERATIONS` template param.

### Ticket

- Closes #21622
- Closes #21890 

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17632788387) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17617996508) CI with demo tests passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/runs/17615116519) CI passes (if applicable)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/runs/17615080611) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

jasondavies committed:
Use exp_init/recip_init instead of hardcoded constants.

This prepares the way for changing these constants in tt-llk.

jasondavies commented:
Passed `./build/test/tt_metal/unit_tests_llk` with my tt-llk changes (`recip` branch).  Not sure how to run sweeps locally though.  Is there a good way for me to check accuracy locally?

```sh
$ python3 tests/sweep_framework/sweeps_runner.py --module-name add
Traceback (most recent call last):
  File "/home/jason/tt-metal/tests/sweep_framework/sweeps_runner.py", line 13, in <module>
    from tt_metal.tools.profiler.process_ops_logs import get_device_data_generate_report
ModuleNotFoundError: No module named 'tt_metal'
```

jasondavies commented:
Looks like I just needed to set `PYTHONPATH` to the tt-metal directory to get Python tests running locally.  I'm looking into the test failures now.

jasondavies committed:
Fix.

jasondavies commented:
I think this should fix the failing tests. I re-ran `tests/ttnn/unit_tests/operations/eltwise/test_binary_bcast.py::test_binary_scalar_ops` locally.

jasondavies committed:
Use exp_init/recip_init instead of hardcoded constants.

(Copied changes from Wormhole.)

jasondavies committed:
Replace sqrt_tile+recip_tile with rsqrt_tile.

jasondavies committed:
Checkpoint.

jasondavies committed:
Update.

philei-tt committed:
Move all profiling dependencies to install_dependencies.sh (#22726)

### Ticket
N/A

### Problem description
Yesterday I tried to compile tt-metal with "--enable-profiler" and got
the following error:
```
../../../server/TracyWorker.cpp:24:10: fatal error: capstone.h: No such file or directory
   24 | #include <capstone.h>
```

I managed to fix that with `sudo apt install libcapstone-dev`

I managed to reproduce that on both ubuntu 24.04 and ubuntu 22.04

### What's changed
Add libcapstone-dev to build dependencies of tt-metal in
./install_dependencies.sh

### Checklist
- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes
- [ ] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml)
CI with demo tests passes (if applicable)
- [ ] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml)
CI passes (if applicable)
- [ ] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml)
CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable) See [recommended dev
flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

pavlejosipovic committed:
Clean up of height sharded conv2d weights kernel reader (#22793)

Clean up of unused runtime args, compile time args and superfluous loops
for height sharded weights reader.

CI is having a bad day, I'll post green pipelines next week.

### Checklist
- [x] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/runs/15348476581)
- [x] [Blackhole nightly
tests](https://github.com/tenstorrent/tt-metal/actions/runs/15375605595)
- [x] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/runs/15375600402)
- [x] [Frequent model and ttnn
tests](https://github.com/tenstorrent/tt-metal/actions/runs/15375596111)
- [x] [Nightly tt-metal L2
tests](https://github.com/tenstorrent/tt-metal/actions/runs/15376458534)

Co-authored-by: Pavle Josipovic <pjosipovic@tenstorrent.com>

abin555 committed:
Check for Duplicate Cores - Issue #6754 (#20907)

### Ticket
[Link to Github
Issue](https://github.com/tenstorrent/tt-metal/issues/6754)

### Problem description
Copied from issue:  
Create a program, add a kernel twice, eg:
```cpp
       auto dm0a = tt_metal::CreateKernel(
            program,
            "path",
            cores,
            tt_metal::DataMovementConfig{
                    .processor = tt_metal::DataMovementProcessor::RISCV_0
            });
       auto dm0b = tt_metal::CreateKernel(
            program,
            "path",
            cores,
            tt_metal::DataMovementConfig{
                    .processor = tt_metal::DataMovementProcessor::RISCV_0
            });
```

The above should error out - creating 2 kernels running on the same
cores should fail.
Test should cover DM and COMPUTE and a variety of overlaps, eg, create
both on a single core, create one on many cores and the other on just of
those, or a set of those, etc.

### What's changed
When adding new kernels, checks for duplicates by iterating through
existing kernels and each used core, checking for an overlap at core
location of kernels of the same type.

### Checklist
- [x] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes
- [ ] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml)
CI passes (if applicable)
- [ ] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml)
CI passes (if applicable)
- [ ] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml)
CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml)
CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

---------

Co-authored-by: Austin Geisert <austin@zutepredictive.com>
Co-authored-by: William Ly <williamly@tenstorrent.com>

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Remove duplicate rsqrt test and restrict inputs to positive values.

This matches the positive values used for the sqrt test.

jasondavies committed:
Fix missing arguments.

jasondavies committed:
Fix breakage caused by duplicate trig code in latest tt-llk.

jasondavies committed:
Fix arguments for Blackhole.

jasondavies committed:
Updates.

jasondavies committed:
Update for Blackhole too.

jasondavies commented:
Note: I had to a remove a failing rsqrt test in `tests/ttnn/unit_tests/operations/eltwise/test_unary.py` (it's tested elsewhere anyway).  This was failing due to not returning +inf in the case of input=0.0 (it returns a very large number instead), which I think is reasonable in these circumstances.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies commented:
This needs integration testing with:

- https://github.com/tenstorrent/tt-llk/pull/230

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies committed:
Fix.

jasondavies committed:
Revert "Fix breakage caused by duplicate trig code in latest tt-llk."

This reverts commit 9d3732804066787eaae0feb47b686694440f2612.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies committed:
Fix.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Restore test_rsqrt now that rsqrt(0.0)=inf.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies commented:
@fvranicTT This requires https://github.com/tenstorrent/tt-llk/pull/230 to run CI -- I think CI still won't work if I use a submodule ref to my branch?  I've run all ttnn unit tests locally on WH and they all pass.  I'm about to run BH locally too.

jasondavies commented:
Waiting on code owner review from @aagarwalTT, @abhullar-tt, @dchenTT, @jbaumanTT, @nhuang-tt, @ntarafdar, @patrickroberts, @sjameelTT, @tt-aho, @tt-asaigal.  Thanks!

jasondavies committed:
Remove 'fast_and_approx' from rsqrt.

It should just be handled the same as sqrt.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies committed:
Remove more fast_and_approx from rsqrt.

jasondavies committed:
Fix another fast_and_approx rsqrt.

jasondavies commented:
I'm re-running WH/BH tests locally.

I found a failure when testing on BH (somehow I missed this on WH): batchnorm, which uses `rsqrt_tile`, which it turns out had a default of `fast_and_approx=true` but confusingly ignored `fast_and_approx`.  With my new rsqrt implementation, it was no longer ignoring `fast_and_approx` but treating it the same as `approx_mode`, so the default had lower precision than expected.

I've made changes so that rsqrt_tile is now treated exactly like sqrt_tile, and respects `approx_mode` instead.  Will update PR description to include this.

jasondavies committed:
Add missing rsqrt include.

jasondavies committed:
Add another missing rsqrt include.

jasondavies committed:
Another rsqrt_tile fix.

jasondavies committed:
Add missing rsqrt include to docs.

jasondavies committed:
Fix.

jasondavies commented:
I suspect the correct resolution here is to just update the test to use TILE layout.  Looking at the history of the test, it was originally added for Grayskull.

I'm not sure why fast_and_approx unary ops are treated differently from unary ops in terms of tensor layout.

jasondavies committed:
Restore eltwise_rsqrt test special handling for RM layout.

jasondavies commented:
Also still need code owner review from @dchenTT, @ntarafdar, @patrickroberts, and/or @sjameelTT.

jasondavies commented:
Actually it looks like a couple of issues persist on BH:

- `models/demos/blackhole/stable_diffusion/tests/test_unet_2d_condition_model.py::test_unet_2d_condition_model_512x512[2-4-64-64-device_params=l1_small_size_24576] - AssertionError: 0.993725881747892`
- `tests/ttnn/unit_tests/operations/fused/test_batch_norm.py::test_batch_norm_fp32[eps=1e-05-bias=False-weight=True-check_mean=True-check_var=True-input_shapes=torch.Size([3, 5, 64, 120])]`

jasondavies commented:
> * tests/ttnn/unit_tests/operations/fused/test_batch_norm.py::test_batch_norm_fp32[eps=1e-05-bias=False-weight=True-check_mean=True-check_var=True-input_shapes=torch.Size([3, 5, 64, 120])]

Looks like this failure is unrelated to my PR?  It fails on BH on latest main: 00d4571afaf6dabdfc51b192b3fc9d86c17df21e at the time of writing.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies commented:
> Looks like this failure is unrelated to my PR? It fails on BH on latest main: [00d4571](https://github.com/tenstorrent/tt-metal/commit/00d4571afaf6dabdfc51b192b3fc9d86c17df21e) at the time of writing.

BH CI is passing on everything except the SD model (unet), so the batch_norm failure must be something to do with my local setup.  Now I just need to work out why the SD model test fails.

jasondavies commented:
> Now I just need to work out why the SD model test fails.

Attempting to debug the slightly low PCC (0.993 versus target 0.996) is quite laborious as each test run of this model takes 5 minutes to complete.  I've tried a number of things:

- test on main branch and replace sqrt+recip with rsqrt - results in a slightly lower PCC
- test on main branch and only update recip or sqrt LLK alone - same issue
- messing with overriding approx_mode and rounding when storing to dst - same issue
- tried different random seeds just in case this generates lower PCCs on the main branch - it doesn't really

Since SD uses bfp8 for several things, I do wonder if there's something accidental going on with the original reciprocal/sqrt creating some kind of bias that "improves" conversion to bfp8 via truncation.  I really have no idea at this point.

jasondavies commented:
The latest CI runs are:

WH: https://github.com/tenstorrent/tt-metal/actions/runs/16140207797
BH: https://github.com/tenstorrent/tt-metal/actions/runs/16140213869

The other issue with the SD model failure is that it only seems to run on BH: the test doesn't run on N300 and I can't see it on N150 on the CI run.  I assume the result would be the same though.

sjameelTT commented on tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py:
```diff
@@ -823,7 +803,7 @@ def test_unary_log_sigmoid_ttnn(input_shapes, device):
 def test_unary_rsqrt_ttnn(input_shapes, device):
-    in_data, input_tensor = data_gen_with_range(input_shapes, -10, 10, device)
+    in_data, input_tensor = data_gen_with_range(input_shapes, 0, 100, device)
     _, output_tensor = data_gen_with_range(input_shapes, -1, 1, device)
```
Any reason why we have a reduced range here?

jasondavies replied:
Square root (and its reciprocal) isn't currently defined for negative inputs.

We could return `nan` at the expense of a small number of additional operations.  There's an open issue in tt-metal about this: https://github.com/tenstorrent/tt-metal/issues/23166 but I suspect it might be better to tackle separately in a new PR.

jasondavies replied:
I should also say: this simply makes the rsqrt range consistent with sqrt, tested elsewhere in the same file.

sjameelTT reviewed (approved):
Apologies for the late review.

sjameelTT commented:
The LLK team might be know a bit more about the PCC drop. @nvelickovicTT maybe you have some insight?

Otherwise @esmalTT thoughts on this PCC drop. Is it substantial for the model?

jasondavies committed:
Adjust PCC.

Ironically this test fails when higher precision sqrt/rsqrt/recip are
used, and it's unclear why, or whether the PCC drop of 0.002 is really
that material.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies commented:
> @jasondavies could this affect op performance? If so, you might want to run the device performance pipeline.

It could be worth doing, though the new LLKs should be both faster and more accurate with the exception of approx_mode sqrt, which takes 13 cycles instead of 5 cycles, but now produces 10 bits of precision.  Replacing sqrt_tile+recip_tile with rsqrt_tile closes this gap of 8 cycles to 3 cycles in the case of approx_mode.  The other LLKs are all fewer cycles compared with the originals.

jasondavies commented:
Including PyTorch's CPU output for comparison:

<img width="512" height="512" alt="output_torch_image" src="https://github.com/user-attachments/assets/c28fec3a-63c4-4f7e-90b2-4b338c42601a" />

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies commented:
Just a brief update on this.

> Note to self: this test failure is much faster to run and similarly shows a slight drop in PCC.
> 
> ```
> pytest 'tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py::test_bert_for_question_answering[num_hidden_layers=1-sequence_size=128-batch_size=8-model_name=mrm8488/bert-tiny-finetuned-squadv2]'
> 
> AssertionError: 0.9552467568064892
> ```
> 
> (target PCC 0.96)

This small drop in PCC can be reproduced by patching ttnn.layer_norm in `models/demos/bert_tiny/tt/bert_tiny.py` so that it uses a PyTorch shim instead:

```py
def my_layer_norm(x, weight=None, bias=None, epsilon=None, memory_config=None):
    device = x.device()
    ref_value = aten.native_layer_norm.default(
            ttnn.to_torch(x),
            [x.shape[-1]],
            ttnn.to_torch(weight[0]),
            ttnn.to_torch(bias[0]),
            epsilon)[0]
    return ttnn.from_torch(
            ref_value,
            layout=ttnn.TILE_LAYOUT,
            device=device,
            memory_config=memory_config,
            dtype=dtype)

ttnn.layer_norm = my_layer_norm
```

This results in a PCC of 0.954473435308662.

At this point, it's difficult to know how to proceed.  The new sqrt/rsqrt/recip functions are all accurate to within 1 ulp.

However, it seems that users of the original functions (specifically layer_norm) have possibly ended up accidentally relying on some aspect of the original functions such as a bias in a particular direction, perhaps combined with truncation/bad rounding elsewhere.  This is just a hypothesis.

bbradelTT commented:
Changing to fp32 reduce in https://github.com/tenstorrent/tt-metal/pull/25259 lead to similar problems with pcc being worse (while ulp was better). I need to figure out why.

Maybe leave layer_norm with the old code, and that code can be updated separately once rsqrt_tile is merged in.

jasondavies commented:
> Maybe leave layer_norm with the old code, and that code can be updated separately once rsqrt_tile is merged in.

Yes, I was wondering about this as a possibility.  It should be possible to do, but seems like a bit of hack.  I'm already testing all possible combinations of sqrt/rsqrt/recip and new_sqrt/new_rsqrt/new_recip in my attempts to figure out the issue.

We could do something like merge this PR with an `old_rsqrt_tile` API for use with layernorm (I think it _should_ be possible to merge the old sqrt+recip and preserve the old behaviour).

Regarding the BERT test that I'm looking at, I'm also not sure why the PCC isn't closer to 1.0.  I wonder if that gives a clue to the real issue elsewhere in that model.

jasondavies committed:
Update tt-llk ref.

bbradelTT commented:
@jasondavies I cloned your branch to https://github.com/tenstorrent/tt-metal/tree/mirror/jasondavies/recip 
updated tt_llk
```
cd tt_metal/third_party/tt_llk
git checkout 77dfc15b5711c83566bf6f17a74dd8222c8c8818
```
and tried running some code. I got the following errors:
```
2025-07-18 19:01:52.206 | critical |          Always | trisc1 build failed. Log: In file included from /localdev/bbradel/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_math_unary_sfpu_api.h:8,
                 from /localdev/bbradel/tt-metal/tt_metal/include/compute_kernel_api.h:22,
                 from /localdev/bbradel/tt-metal/tt_metal/include/compute_kernel_api/layernorm.h:7,
                 from /localdev/bbradel/tt-metal/ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm.cpp:16,
                 from ../chlkc_math.cpp:3,
                 from /localdev/bbradel/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/common/chlkc_list.h:21,
                 from /localdev/bbradel/tt-metal/tt_metal/hw/firmware/src/trisck.cc:14:
/localdev/bbradel/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_abs.h:8:10: fatal error: llk_math_eltwise_unary_sfpu_params.h: No such file or directory
    8 | #include "llk_math_eltwise_unary_sfpu_params.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
 (assert.hpp:107)
2025-07-18 19:01:52.456 | critical |          Always | trisc2 build failed. Log: In file included from /localdev/bbradel/tt-metal/tt_metal/include/compute_kernel_api/common_globals.h:24,
                 from /localdev/bbradel/tt-metal/tt_metal/include/compute_kernel_api/common.h:7,
                 from /localdev/bbradel/tt-metal/tt_metal/include/compute_kernel_api/reduce.h:7,
                 from /localdev/bbradel/tt-metal/ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/compute/layernorm.cpp:13,
                 from ../chlkc_pack.cpp:3,
                 from /localdev/bbradel/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/common/chlkc_list.h:29,
                 from /localdev/bbradel/tt-metal/tt_metal/hw/firmware/src/trisck.cc:14:
/localdev/bbradel/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_pack_api.h: In function 'void llk_pack_fast_tilize_hw_configure(const llk_pack_params_t*)':
/localdev/bbradel/tt-metal/tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_pack_api.h:294:5: error: '_llk_pack_fast_tilize_hw_configure_' was not declared in this scope; did you mean 'llk_pack_fast_tilize_hw_configure'?
  294 |     _llk_pack_fast_tilize_hw_configure_<is_fp32_dest_acc_en>(pack_src_format[output_id], pack_dst_format[output_id]);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     llk_pack_fast_tilize_hw_configure
```

Then I tried the clean / clear and rebuild and got
```
-- Installing: /localdev/bbradel/tt-metal/build_Debug/libexec/tt-metalium/tt_metal/third_party/tt_llk/tt_llk_blackhole/common/inc/sfpu/ckernel_sfpu_abs.h
CMake Error at tt_metal/cmake_install.cmake:424 (file):
  file INSTALL cannot find
  "/localdev/bbradel/tt-metal/tt_metal/third_party/tt_llk/tt_llk_blackhole/common/inc/sfpu/ckernel_sfpu_activations.h":
  No such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:62 (include)
```

Not sure what went wrong.

I think these changes will need to be merged in and then layernorm etc. updated in separate PRs.

bbradelTT commented:
Thanks @jasondavies
Your push fixed things. I created mirror/jasondavies/recip2 and things worked.

jasondavies committed:
Update tt-llk.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Use new layernorm_compat template param for rsqrt.

jasondavies committed:
Checkpoint.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Tweak.

jasondavies committed:
Use compat mode for groupnorm.

jasondavies committed:
Update tt-llk.

jasondavies committed:
Update tt-llk.

jasondavies committed:
Add missing fp32_dest_acc_en template params for BH.

jasondavies committed:
Add missing fp32_dest_acc_en param for BH.

jasondavies commented:
Looks like a slight PCC wobble here:

`FAILED models/demos/vit/tests/pcc/test_ttnn_optimized_sharded_vit_wh.py::test_vit[224-3-224-8-google/vit-base-patch16-224] - AssertionError: 0.8773253211253444`

Target PCC is 0.88.

Will look into whether there's another kernel that needs `layernorm_compat=true`.

jasondavies committed:
Reduce PCC threshold slightly.

jasondavies commented:
> `FAILED models/demos/vit/tests/pcc/test_ttnn_optimized_sharded_vit_wh.py::test_vit[224-3-224-8-google/vit-base-patch16-224] - AssertionError: 0.8773253211253444`

I've adjusted PCC from 0.88 to 0.87 as the change is only marginal, and the PCC is already pretty low for this test.  Similar to the other issues encountered, I think they all related to a sum reduction being performed over relatively low precision values.  My hypothesis remains that sum reduction is the real culprit here and the use of high-precision sqrt/recip simply exposes the underlying issue.

I should also note that I previous did some tests where I inflated the values by 5-10% and this actually improved the PCC slightly, which may support the sum reduction hypothesis.

In this case, it's worth noting that the PCC was marginally improved by setting approx_mode=True, but it was still fractionally below the 0.88 target.  (Funnily enough, if a further change is made to use truncation instead of rounding, the 0.88 target can _just about_ be achieved.)  I think it makes sense to get these changes through (assuming all other intensive tests pass) and tackle the various problems with sum reduction elsewhere.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

fvranicTT commented:
Running tests...
APC: https://github.com/tenstorrent/tt-metal/actions/runs/16678405401
BPC: https://github.com/tenstorrent/tt-metal/actions/runs/16678408424

jasondavies commented:
Other tests are in progress, but here's one failure: https://github.com/tenstorrent/tt-metal/actions/runs/16678405401/job/47212151594

```
2025-08-01 15:39:50.031 | INFO     | models.demos.llama3_70b_galaxy.demo.text_demo:test_demo_text:745 - PCC: 0.9106052269372641
2025-08-01 15:39:50.031 | INFO     | models.demos.llama3_70b_galaxy.demo.text_demo:test_demo_text:746 - Teacher forced token at prefill FAILED PCC check with torch reference model
Error: test_demo_text[wormhole_b0-mesh_device0-device_params0-10-performance-apc-test]

AssertionError: Prefill PCC check failed: 0.9106052269372641, while expected 0.9132907037883032.
  If it is expected to be different in Llama model, please update the text_demo_targets.json file.
  See the comment on the text_demo.py by the assert for instructions.
assert 0.9106052269372641 == 0.9132907037883032
FAILED
```

jasondavies commented:
Another: https://github.com/tenstorrent/tt-metal/actions/runs/16671165150/job/47212316902?pr=21623

```
FAILED tests/ttnn/unit_tests/operations/test_bernoulli.py::test_bernoulli_p_zero[out_dtype=bfloat16-in_dtype=float32-shape=[1, 512, 2, 256]] - AssertionError: Output must be all zeros when p=0
```

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Update tt-llk.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies committed:
Update _sfpu_reciprocal_ params.

jasondavies committed:
Set layernorm_compat = true by default.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies committed:
Update tt-llk.

jasondavies commented:
@bbradelTT If someone can run CI on this latest branch, that would be great.  I've now set layernorm_compat=true by default and only disabled it for ttnn.rsqrt.

However, it's possible that failures may still occur if `recip_tile` or `sqrt_tile` are used alone in combination with reduce sum, as `layernorm_compat` only applies to `rsqrt_tile`.

Of course, I could add more compat flags, to recip_tile and sqrt_tile if that's easier.

FYI, I also took the time to update the LLK to use the new SFPARECIP for BH, and I rewrote the WH reciprocal function completely to shave off another two cycles, along with more robust handling of ±0, ±inf.

Reciprocal now achieves maximum accuracy for all ±0, ±inf, and normal numbers (with the exception of ±8.50705917302346158658e+37 on BH, where we get ±0 instead of ±1.1754943508222875e-38).

jasondavies committed:
Set layernorm_compat=false for ttnn.rsqrt.

jasondavies commented:
> disabled it for ttnn.rsqrt

I hadn't set layernorm_compat=false for ttnn.rsqrt, that's probably a big reason for these CI failures.  Let's retry please.  Thanks!

/cc @fvranicTT

fvranicTT commented:
APC: https://github.com/tenstorrent/tt-metal/actions/runs/17140033964
BPC: https://github.com/tenstorrent/tt-metal/actions/runs/17140044549

FYI @jasondavies

jasondavies committed:
Add legacy_compat to sqrt_tile and recip_tile.

Rename layernorm_compat to legacy_compat.

For now, the defaults are:
- rsqrt_tile: false
- sqrt_tile: false
- recip_tile: true

jasondavies committed:
Update tt-llk.

jasondavies committed:
Checkpoint.

- Restores PCC target 0.88 for models/demos/vit/tests/pcc/test_ttnn_optimized_sharded_vit_wh.py::test_vit[224-3-224-8-google/vit-base-patch16-224].
- ttnn.{rsqrt,sqrt,recip} now set legacy_compat=false.
- rsqrt_tile: legacy_compat=false by default.
- sqrt_tile/recip_tile: legacy_compat=true by default.

jasondavies committed:
Update tt-llk.

jasondavies commented:
@fvranicTT Let's try again, please.

I went ahead and added legacy_compat modes to sqrt_tile and recip_tile, default true.  I restored the original PCC of 0.88 to the ViT test and it passes for me now.

As for ttnn.{rsqrt,sqrt,recip}, I've set legacy_compat=false.  And rsqrt_tile has legacy_compat=false.  So we're mostly going to be in legacy_compat mode in terms of kernels that don't override the defaults.

philei-tt commented:
@jasondavies I managed to fix broken stable diffusion tests from here https://github.com/tenstorrent/tt-metal/pull/26258 by changing how binary ops pass scalars to kernel: https://github.com/tenstorrent/tt-metal/pull/27171/commits/bab87a7ddf8c58e5c40031673f19478f7c5a2cbd Apparently when we use tie-to-even for fp32->bf16, accuracy of some binary ops, like ttnn.mul, can get worse?

jasondavies commented:
@philei-tt Interesting, thanks.  This may well be related to some of failures I'm seeing with my improved LLKs.  I think I'll continue with the legacy_compat approach for now, and it can be investigated more thoroughly later by toggling legacy_compat.

jasondavies committed:
Merge remote-tracking branch 'upstream/main' into recip

jasondavies committed:
Update tt-llk.

jasondavies committed:
Replace sqrt+recip with rsqrt.

jasondavies committed:
Use legacy_compat here for consistency.

jasondavies committed:
Fix missing template param.

jasondavies commented:
@fvranicTT Should be good to give it another CI run now, thanks.

nmauriceTT commented:
@jasondavies I've run the CI.

APC: https://github.com/tenstorrent/tt-metal/actions/runs/17263943267
BH: https://github.com/tenstorrent/tt-metal/actions/runs/17263956889

jasondavies commented:
Looks like BPC is passing. APC has two failures but one is a timeout and the other seems to be unrelated.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies committed:
Fix duplicate case value 'RELU' accidentally added in merge.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

jasondavies commented:
@nmauriceTT Just following up on the results of your model-related pipelines, where we have a couple of failures.  I haven't had a chance to really look at them in-depth and figure out the root cause.

I also just wanted to clarify the current defaults:

- recip_tile: legacy_compat=true
- sqrt_tile: legacy_compat=true
- rsqrt_tile: legacy_compat=false (this is because the original rsqrt_tile used a high number of 25 Newton-Raphson iterations by default (!) so high-precision should be fine)

The fact that rsqrt_tile has a different default means that where I've replaced recip_tile+sqrt_tile with rsqrt_tile, I've had to explicitly set legacy_compat=true.

Further, I made `ttnn.{recip,sqrt,rsqrt}` use `{recip,sqrt,rsqrt}_tile<legacy_compat=false>`.  It's this latter case that could have caused issues if some models were using ttnn and relying on the old behaviour.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

bbradelTT commented:
Another area where more accuracy reduces pcc is rounding. There had to be explicit truncates placed in certain places to get the old behaviour back. E.g. https://github.com/tenstorrent/tt-metal/pull/27911

jasondavies committed:
Update tt-llk.

fvranicTT commented:
BH post-commit is consistently failing with this change. We can’t merge it as is.
Please consider our next steps. FYI @bbradelTT @nmauriceTT @nvelickovicTT @amahmudTT

jasondavies commented:
Doing a build now to try and reproduce.  The two failures that I see are:

```
FAILED tests/ttnn/unit_tests/operations/data_movement/test_slice_write.py::test_slice_write_nd[layout=Layout.ROW_MAJOR-rank=6] - AssertionError: 0.6874356787473929
```

This seems unrelated to my changes and the other changes since the last tt-llk uplift?

```
FAILED tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_eltwise_unary.py::TestEltwiseUnary::test_run_eltwise_atan_op[atan-output_mem_config0-input_mem_config0-input_shapes1] - AssertionError: eltwise-atan test failed with input shape [[1, 1, 320, 384]]. Max ATOL Delta: 6.192449487634432e+16, Max RTOL Delta: 1.3671875, PCC: 0.00271462568066949, PCC check failed
```

This one looks more like it could be something.

jasondavies commented:
To give a flavour:

Input:

```
[tensor([[[[ 473088.,  509952.,  905216.,  ...,  651264., -897024., -917504.],                                                                                    [-179200., -983040., -720896.,  ...,  659456., -446464.,  186368.],
          [ 391168.,  991232., -135168.,  ...,  786432.,  696320.,  419840.],                                                                                     ...,
          [-214016.,  528384., -401408.,  ..., -602112.,  -72192., -557056.],                                                                                     [-356352.,  -40192., -655360.,  ..., -231424.,  462848., -708608.],
          [ 618496., -655360.,  362496.,  ..., -205824., -232448., -278528.]]]],
       dtype=torch.bfloat16)]
```

TTNN:

```
tensor([[[[ 1.5703,  1.5625,  1.5625,  ...,  1.5625, -1.5625, -1.5625],                                                                                           [-1.5625, -1.5625, -1.5625,  ...,  1.5625, -1.5625,  1.5625],
          [ 1.5625,  1.5625, -1.5625,  ...,  1.5625,  1.5625,  1.5625],
          ...,
          [-1.5625,  1.5625, -1.5625,  ..., -1.5625, -1.5625, -1.5625],
          ...,
          [-1.5625,  1.5625, -1.5625,  ..., -1.5625, -1.5625, -1.5625],
          [-1.5625, -1.5469, -1.5625,  ..., -1.5625,  1.5625, -1.5625],
          [ 1.5625, -1.5625,  1.5625,  ..., -1.5625, -1.5625, -1.5625]]]],
       dtype=torch.bfloat16)

```

Torch:

```  
tensor([[[[ 1.5703,  1.5703,  1.5703,  ...,  1.5703, -1.5703, -1.5703],
          [-1.5703, -1.5703, -1.5703,  ...,  1.5703, -1.5703,  1.5703],
          [ 1.5703,  1.5703, -1.5703,  ...,  1.5703,  1.5703,  1.5703],
          ...,
          [-1.5703,  1.5703, -1.5703,  ..., -1.5703, -1.5703, -1.5703],
          [-1.5703, -1.5703, -1.5703,  ..., -1.5703,  1.5703, -1.5703],
          [ 1.5703, -1.5703,  1.5703,  ..., -1.5703, -1.5703, -1.5703]]]],
       dtype=torch.bfloat16)
```

So the LSB is different after the first value.  After `tt-smi -r 0`, it passes with all values `±1.5703`.

nmauriceTT commented:
Yeah, with an input range on [-1e6; 1e6], the values should (virtually) be -pi/2 or pi/2 virtually all the time.
From what I see, the differences we have  compared to torch are tiny (1-2 ULPs at most). This is typically a case where we should probably rely on assert_with_ulp or assert_allclose rather than PCC.

What's interesting is that the test case only seem to be failing on blackhole. It does not seem to be a problem on wormholes. https://github.com/tenstorrent/tt-metal/actions/runs/17505018481 

The fact that the LSB stops being adjusting after reseting the board is also puzzling. It smells like some stochastic rounding is happening (but to be confirmed).

jasondavies commented:
I have found the root cause: on BH, the constant 2.0 was only initialised for reciprocal for approximate_mode=false (it's not needed otherwise).

But sfpu_reciprocal is called without template params by atan, which defaults to non-approximate mode, even though atan itself has a notion of approximate mode, and calls recip_init<APPROXIMATION_MODE>.

On WH, we always initialise constants for reciprocal as we also need them for approximate mode.

Two options:

1. Change atan to use sfpu_reciprocal<max_iter=0> in the case of approximate mode.  I haven't done the analysis but I believe this makes sense and the output precision should be roughly the same as the precision of the approximate reciprocal (in fact the full polynomial is probably overkill for approximate mode).
2. Call recip_init<false> instead, which aligns with the use of sfpu_reciprocal<max_iter=2>.

Even though the first option is probably the more "correct" approach in line with the meaning of approximate mode, it may make more sense to do the second option initially to get the PR merged as it minimises potential downstream effects of changing atan to be less precise for approximate mode.

jasondavies committed:
Fix sfpu_reciprocal initialisation.

jasondavies commented:
@nmauriceTT @fvranicTT Just pushed the initialisation fix.

I did some minor cleanup to atan, e.g. removing unnecessary load/store via `dst_reg[0]` and using `setsgn` instead of an unnecessary conditional.  It's possible the strange use of `dst_reg[0]` was originally done to reduce register pressure, but it's not documented as such and I can't see this function being inlined anywhere that would indicate this was necessary.

jasondavies committed:
Simplify a bit to align with BH.

jasondavies committed:
Use sfpu_reciprocal instead of _sfpu_reciprocal_ in softsign.

nmauriceTT commented:
@jasondavies Thank you for the fix and clean up. 

I've re-launched CI:
APC: https://github.com/tenstorrent/tt-metal/actions/runs/17526143527
BH APC: https://github.com/tenstorrent/tt-metal/actions/runs/17526135486

jasondavies commented:
@nmauriceTT Thanks.

- BPC is now passing except for the `test_slice_write.py::test_slice_write_nd` that occurred last time and is unrelated.
- APC still waiting for ttnn-unit-tests N300 but looks promising so far, except for failing t3k fabric tests but I'm assuming these are also unrelated.

PR Gate is failing due to missing auth token issue with ttsim-integration.

jasondavies commented:
- APC passing except for t3k fabric tests (unrelated?) and a timeout issue with one worker.

jasondavies committed:
Merge commit '15cac5ab0c9093c35bc97a464431481cb20fb5d3' into recip

jasondavies commented:
@nmauriceTT @fvranicTT Merged with main (up to ttsim fix).  Should fix the failing tools test as well as the ttsim error.

jasondavies commented:
@nmauriceTT I guess I could attempt a fix for the ttsim issue in this PR?  I think this might be the preferred way:

```diff
--- a/.github/workflows/ttsim.yaml
+++ b/.github/workflows/ttsim.yaml
@@ -15,7 +15,7 @@ on:

 jobs:
   ttsim-integration:
-    if: github.repository == 'tenstorrent/tt-metal'
+    if: github.event.pull_request.head.repo.full_name == github.repository
     runs-on: ubuntu-latest
     permissions:
       contents: read
```

fvranicTT commented:
> ```diff
> --- a/.github/workflows/ttsim.yaml
> +++ b/.github/workflows/ttsim.yaml
> @@ -15,7 +15,7 @@ on:
> 
>  jobs:
>    ttsim-integration:
> -    if: github.repository == 'tenstorrent/tt-metal'
> +    if: github.event.pull_request.head.repo.full_name == github.repository
>      runs-on: ubuntu-latest
>      permissions:
>        contents: read
> ```

Hi @jasondavies, with this change, this job will only run if the pull request comes from a branch inside the same repository (e.g. `tenstorrent/tt-metal:feature-x`), and will be skipped for forks.
Still, changes coming from forks could break ttsim and we wouldn’t notice until the next PR workflow is executed. The root issue is that PRs from forks don’t have access to repository secrets. We’ll need to find a way to handle that.
FYI @blozano-tt @afuller-TT

jasondavies commented:
Hey @fvranicTT

> this job will only run if the pull request comes from a branch inside the same repository (e.g. `tenstorrent/tt-metal:feature-x`), and will be skipped for forks.

Yes, that was the stated intention of @blozano-tt's https://github.com/tenstorrent/tt-metal/pull/28133 but it failed to work as expected.  My snippet above achieves the stated aim.

As you say, maybe the root issue should be solved a different way.

jasondavies commented:
> I made a copy of the PR: #28180

I forgot this means we have to wait for approval from all the code owners again.  At least now we know ttsim-integration _does_ pass.

jasondavies commented:
Model perf failure:

> FAILED models/experimental/efficientnetb0/tests/perf/test_e2e_performant.py::test_e2e_performant_dp[wormhole_b0-resolution0-1-act_dtype0-weight_dtype0-device_params0] - AssertionError: 0.9095259619248233

This is failing a PCC check of 0.91.

jasondavies committed:
Skip ttsim-integration for PRs from forks.

Testing for @blozano-tt

jasondavies committed:
Adjust vovnet expected perf.

jasondavies committed:
ttsim: trying alternative check for fork PR.

jasondavies commented:
> Model perf failure:
> 
> > FAILED models/experimental/efficientnetb0/tests/perf/test_e2e_performant.py::test_e2e_performant_dp[wormhole_b0-resolution0-1-act_dtype0-weight_dtype0-device_params0] - AssertionError: 0.9095259619248233
> 
> This is failing a PCC check of 0.91.

This can be worked around by making the following change:

```diff
--- a/models/experimental/efficientnetb0/tt/efficientnetb0.py
+++ b/models/experimental/efficientnetb0/tt/efficientnetb0.py
@@ -451,7 +451,7 @@ class Efficientnetb0:
         x = self._blocks15(x)
         x = self._conv_head(x)

-        x = x * ttnn.sigmoid(x)
+        x = x * ttnn.sigmoid_accurate(x)

         x = ttnn.sharded_to_interleaved(x, memory_config=ttnn.L1_MEMORY_CONFIG)
         x = ttnn.to_layout(x, layout=ttnn.ROW_MAJOR_LAYOUT)
```

jasondavies commented:
/cc @nmauriceTT I know there's an open issue about sigmoid accuracy https://github.com/tenstorrent/tt-metal/issues/26946 - does the above change to efficientnetb0 seem reasonable in the circumstances?

jasondavies committed:
Revert ttsim changes.

jasondavies committed:
Use sigmoid_accurate instead of sigmoid.

This results in a small increase in PCC, which is enough to offset a
small drop in PCC caused by improved reciprocal accuracy.

jasondavies commented:
> Let's check perf, too. Model perf regression: https://github.com/tenstorrent/tt-metal/actions/runs/17586418067 Device perf tests: https://github.com/tenstorrent/tt-metal/actions/runs/17586415052

Let's re-run these if possible.

@ssinghalTT Can you review the vovnet perf update?

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip

nmauriceTT commented:
> /cc @nmauriceTT I know there's an open issue about sigmoid accuracy https://github.com/tenstorrent/tt-metal/issues/26946 - does the above change to efficientnetb0 seem reasonable in the circumstances?

In this context, I think it's reasonable. Ideally, we want to know why the tests are failing in the first place. I think it may be a similar situation as with LayerNorm/GroupNorm/Softmax, where computing the mean/variance/sum was giving inaccurate results but that were getting 'negated' by the 'errors' in current reciprocate/sqrt/...

fvranicTT commented:
> There is a new failure (thanks to merging with latest main) relating to `tests/ttnn/unit_tests/operations/fused/test_layer_norm.py::test_large_layer_norm` which was updated recently #28190, but it fails on main too.

Yeah, we saw that too. Working on finding someone to fix this.

bbradelTT commented:
> > There is a new failure (thanks to merging with latest main) relating to `tests/ttnn/unit_tests/operations/fused/test_layer_norm.py::test_large_layer_norm` which was updated recently #28190, but it fails on main too.
> 
> Yeah, we saw that too. Working on finding someone to fix this.

Sorry. That was from someone on my team. APC passed on their branch, including that specific test. A fix PR should be up in a few minutes.

jasondavies committed:
Merge remote-tracking branch 'origin/main' into recip
